src/main/java/org/jboss/netty/handler/execution/FairOrderedMemoryAwareThreadPoolExecutor.java

```diff
@@ -0,0 +1,259 @@
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.jboss.netty.handler.execution;
+
+import org.jboss.netty.channel.Channel;
+import org.jboss.netty.channel.ChannelEvent;
+import org.jboss.netty.channel.ChannelState;
+import org.jboss.netty.channel.ChannelStateEvent;
+import org.jboss.netty.util.ObjectSizeEstimator;
+import org.jboss.netty.util.internal.ConcurrentIdentityWeakKeyHashMap;
+
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+/**
+ * <pre>
+ * This is a <b>fair</b> alternative of {@link OrderedMemoryAwareThreadPoolExecutor} .
+ *
+ * <h3>Unfair of {@link OrderedMemoryAwareThreadPoolExecutor}</h3>
+ * The task executed in {@link OrderedMemoryAwareThreadPoolExecutor} is unfair in some situations.
+ * For example, let's say there is only one executor thread that handle the events from the two channels, and events
+ * are submitted in sequence:
+ *
+ *           Channel A (Event A1) , Channel B (Event B), Channel A (Event A2) , ... , Channel A (Event An)
+ *
+ * Then the events maybe executed in this unfair order:
+ *
+ *          ----------------------------------------&gt; Timeline --------------------------------&gt;
+ *           Channel A (Event A1) , Channel A (Event A2) , ... , Channel A (Event An), Channel B (Event B)
+ *
+ * As we see above, Channel B (Event B) maybe executed unfairly late.
+ * Even more, if there are too much events come in Channel A, and one-by-one closely, then Channel B (Event B) would be
+ * waiting for a long while and become "hungry".
+ *
+ * <h3>Fair of FairOrderedMemoryAwareThreadPoolExecutor</h3>
+ * In the same case above ( one executor thread and two channels ) , this implement will guarantee execution order as:
+ *
+ *          ----------------------------------------&gt; Timeline --------------------------------&gt;
+ *           Channel A (Event A1) , Channel B (Event B), Channel A (Event A2) , ... , Channel A (Event An),
+ *
+ * </pre>
+ * <b>NOTE</b>: For convenience the case above use <b>one single executor thread</b>, but the fair mechanism is suitable
+ * for <b>multiple executor threads</b> situations.
+ */
+public class FairOrderedMemoryAwareThreadPoolExecutor extends MemoryAwareThreadPoolExecutor {
+
+    // end sign
+    private final EventTask end = new EventTask(null);
+
+    private final AtomicReferenceFieldUpdater<EventTask, EventTask> fieldUpdater =
```

Comment:
Cannot be static, because "EventTask" is a member inner class ( it cannot be static inner class, since it uses method "doUnorderedExecute" of parent class )